clusterip service not work after running big service(1000 endpoints) with antrea-proxy enabled along with kube-proxy

**Describe the bug**
I enabled antrea-proxy along with kube-proxy in our k8s clusters and create big services with 250 endpoints, 1000 endpoints and 5000 endpoints. After the testing, the clusterip service functionality is broken. 

**To Reproduce**
1. create 500 worker nodes cluster in TKGM1.3 with antrea-proxy enabled
2. download clusterloader2 source code and run the scripts of dep_svc_l.yaml, dep_svc_xl.yaml and dep_svc_exl.yaml by sequential. 

**Expected**
clusterip service still works.

**Actual behavior**
after the testing, I found that the grafana can not show the dashboard, due to grafana failed to  access the prometheus clusterip service. The clusterip service function is broken.
I can succeed to access the pod's ip but failed to curl its service ip.

```
root@systest-runner:/tmp[40430]# kubectl  get pods -n monitoring -o wide
NAME                                  READY   STATUS    RESTARTS   AGE   IP             NODE                                NOMINATED NODE   READINESS GATES
grafana-69c5568879-hd2sm              1/1     Running   0          18h   100.96.213.3   workload-21-md-0-7cf7d99f68-snf2m   <none>           <none>
prometheus-k8s-0                      3/3     Running   1          18h   100.96.17.3    workload-21-md-0-7cf7d99f68-57jcs   <none>           <none>
prometheus-operator-7bf997fc6-vlk89   1/1     Running   0          18h   100.97.179.3   workload-21-md-0-7cf7d99f68-bxb6b   <none>           <none>
root@systest-runner:/tmp[40430]# kubectl  get svc -n monitoring
NAME                  TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                       AGE
grafana               NodePort    100.64.30.207    <none>        3000:32066/TCP                18h
master                ClusterIP   None             <none>        6443/TCP,2379/TCP,10250/TCP   18h
prometheus-k8s        NodePort    100.66.117.106   <none>        9090:31820/TCP                18h
prometheus-operated   ClusterIP   None             <none>        9090/TCP                      18h
prometheus-operator   ClusterIP   None             <none>        8080/TCP                      18h

/ $ curl http://100.96.17.3:9090
<a href="/graph">Found</a>.

/ $ curl http://100.66.117.106:9090
curl: (28) Failed to connect to 100.66.117.106 port 9090: Operation timed out
```

 #**Versions:**
Please provide the following information:
 - Antrea version (Docker image tag). **v0.11.3**
 - Kubernetes version (use `kubectl version`). If your Kubernetes components have [different versions] **v1.20.4**
 (https://kubernetes.io/docs/setup/release/version-skew-policy/), please provide the version for all of them.
 - Container runtime: which runtime are you using (e.g. containerd, cri-o, docker) and which version are you using? containerd
 - Linux kernel version on the Kubernetes Nodes (`uname -r`). **5.4.0-66-generic**
 - If you chose to compile the Open vSwitch kernel module manually instead of using the kernel module built into the Linux kernel, which version of the OVS kernel module are you using? Include the output of `modinfo openvswitch` for the Kubernetes Nodes.

**Additional context**
Add any other context about the problem here, such as Antrea logs, kubelet logs, etc.
```
kubectl  exec -it antrea-agent-2r7rt -n kube-system -c antrea-ovs --  ovs-ofctl dump-groups br-int
NXST_GROUP_DESC reply (xid=0x2):
kubectl  exec -it antrea-agent-2r7rt -n kube-system -c antrea-ovs -- ovs-ofctl dump-flows br-int --no-stats --names > out2.log
```

 Checking the antrea error log in the /var/log/antrea in each worker node, we could see the errors as following
```
E0413 15:30:09.010772       1 proxier.go:280] Error when installing Endpoints groups: error when installing Service Endpoints Group: message is canceled because of disconnection from the Switch
E0413 15:30:09.012386       1 proxier.go:275] Error when installing Endpoints flows: message is canceled because of disconnection from the Switch
E0413 15:30:10.077826       1 client.go:670] Error when replaying cached group 5: message is canceled because of disconnection from the Switch
E0413 15:30:10.077861       1 client.go:670] Error when replaying cached group 1: message is canceled because of disconnection from the Switch
E0413 15:30:10.077871       1 client.go:670] Error when replaying cached group 8: message is canceled because of disconnection from the Switch
E0413 15:30:10.077879       1 client.go:670] Error when replaying cached group 7: message is canceled because of disconnection from the Switch
E0413 15:30:10.077976       1 client.go:670] Error when replaying cached group 27: message is canceled because of disconnection from the Switch
E0413 15:30:10.077985       1 client.go:670] Error when replaying cached group 9: message is canceled because of disconnection from the Switch
E0413 15:30:10.077994       1 client.go:670] Error when replaying cached group 3: message is canceled because of disconnection from the Switch
E0413 15:30:10.078001       1 client.go:670] Error when replaying cached group 6: message is canceled because of disconnection from the Switch
E0413 15:30:10.078008       1 client.go:670] Error when replaying cached group 2: message is canceled because of disconnection from the Switch
E0413 15:30:12.819763       1 client.go:663] Error when replaying cached flows: one message in bundle failed
E0413 15:30:13.342754       1 client.go:663] Error when replaying cached flows: one message in bundle failed
E0413 15:30:13.439770       1 client.go:663] Error when replaying cached flows: one message in bundle failed
E0413 15:30:13.681335       1 client.go:663] Error when replaying cached flows: one message in bundle failed
E0413 15:30:14.100804       1 client.go:663] Error when replaying cached flows: one message in bundle failed
E0413 15:30:14.453300       1 client.go:663] Error when replaying cached flows: one message in bundle failed
E0413 15:30:14.604515       1 client.go:663] Error when replaying cached flows: one message in bundle failed
E0413 15:30:14.731602       1 client.go:663] Error when replaying cached flows: one message in bundle failed
E0413 15:30:14.770858       1 client.go:663] Error when replaying cached flows: one message in bundle failed
```
Whole bundle of test cases and log refer to tarball.
[log.tar.gz](https://github.com/vmware-tanzu/antrea/files/6309131/log.tar.gz)


(Please consider pasting long output into a [GitHub gist](https://gist.github.com) or any other pastebin.)

Unfortunately, guys, I'm stuck with bundle issue.  There are 2 issues:
1) I didn't find the reason that why when added a group in bundle,  deleting the group with groupID failed, but deleting it in bundle way can work.
2) INSERT-BUCKET doesn't work as expected. I write a simple test in PR #2101 https://github.com/vmware-tanzu/antrea/blob/b320a4dd63b8b51f9510ff74b1734897ae58663e/test/integration/agent/openflow_test.go#L487, the test has 4 endpoint, and I set the max number of inserting buckets is 3. The last one should be added by INSERT-BUCKET, but it failed. The output is below:
```
=== RUN   TestGroup
I0416 18:35:50.426456    2012 entry.go:314] Initialize connection or re-connect to /var/run/openvswitch/br01.mgmt.
I0416 18:35:51.427242    2012 entry.go:314] Connected to socket /var/run/openvswitch/br01.mgmt
I0416 18:35:51.427587    2012 entry.go:359] New connection..
I0416 18:35:51.427848    2012 entry.go:314] Send hello with OF version: 4
I0416 18:35:51.427953    2012 entry.go:359] Received Openflow 1.3 Hello message
I0416 18:35:51.428613    2012 entry.go:314] Received ofp1.3 Switch feature response: {Header:{Version:4 Type:6 Length:32 Xid:3} DPID:00:00:f6:81:90:ef:37:48 Buffers:0 NumTables:254 AuxilaryId:0 pad:[0 0] Capabilities:79 Actions:0 Ports:[]}
I0416 18:35:51.428717    2012 entry.go:359] Openflow Connection for new switch: 00:00:f6:81:90:ef:37:48
I0416 18:35:51.429269    2012 ofctrl_bridge.go:221] OFSwitch is connected: 00:00:f6:81:90:ef:37:48
E0416 18:35:51.433889    2012 entry.go:314] Received Vendor error: OFPBFC_MSG_FAILED on ONFT_BUNDLE_CONTROL message
    openflow_test.go:543:
                Error Trace:    openflow_test.go:543
                Error:          Received unexpected error:
                                error when installing Service Endpoints Group: one message in bundle failed
                Test:           TestGroup
                Messages:       no error should return when installing groups for Service
--- FAIL: TestGroup (1.02s)
FAIL
E0416 18:35:51.434699    2012 entry.go:314] Received Vendor error: NXBRC_UNENCODABLE_ERROR on ONFT_BUNDLE_ADD_MESSAGE message
I0416 18:35:51.434753    2012 entry.go:359] Controller is delete
coverage: 3.1% of statements in github.com/vmware-tanzu/antrea/pkg/...
FAIL    github.com/vmware-tanzu/antrea/test/integration/agent   1.062s
testing: warning: no tests to run
PASS
coverage: 1.2% of statements in github.com/vmware-tanzu/antrea/pkg/...
ok      github.com/vmware-tanzu/antrea/test/integration/ovs     0.043s  coverage: 1.2% of statements in github.com/vmware-tanzu/antrea/pkg/... [no tests to run]
FAIL
```

Could anyone give me some suggestions? Thanks.

--- a/docs/feature-gates.md
+++ b/docs/feature-gates.md
@@ -51,7 +51,10 @@ example, to enable `AntreaProxy` on Linux, edit the Agent configuration in the
 `AntreaProxy` implements Service load-balancing for ClusterIP Services as part
 of the OVS pipeline, as opposed to relying on kube-proxy. This only applies to
 traffic originating from Pods, and destined to ClusterIP Services. In
-particular, it does not apply to NodePort Services.
+particular, it does not apply to NodePort Services. Kindly remind that due
+to some restrictions on the implementation of Services in Antrea, the maximum
+number of Endpoints that Antrea can support at the moment is 800. If the number
+of Endpoints for a given Service exceeds 800, extra Endpoints will be dropped.
 
 Note that this feature must be enabled for Windows. The Antrea Windows YAML
 manifest provided as part of releases enables this feature by default. If you

--- a/docs/feature-gates.md
+++ b/docs/feature-gates.md
@@ -51,10 +51,11 @@ example, to enable `AntreaProxy` on Linux, edit the Agent configuration in the
 `AntreaProxy` implements Service load-balancing for ClusterIP Services as part
 of the OVS pipeline, as opposed to relying on kube-proxy. This only applies to
 traffic originating from Pods, and destined to ClusterIP Services. In
-particular, it does not apply to NodePort Services. Kindly remind that due
-to some restrictions on the implementation of Services in Antrea, the maximum
-number of Endpoints that Antrea can support at the moment is 800. If the number
-of Endpoints for a given Service exceeds 800, extra Endpoints will be dropped.
+particular, it does not apply to NodePort Services. Please note that due to
+some restrictions on the implementation of Services in Antrea, the maximum
+number of Endpoints that Antrea can support at the moment is 800. If the
+number of Endpoints for a given Service exceeds 800, extra Endpoints will
+be dropped.
 
 Note that this feature must be enabled for Windows. The Antrea Windows YAML
 manifest provided as part of releases enables this feature by default. If you

--- a/pkg/agent/openflow/client.go
+++ b/pkg/agent/openflow/client.go
@@ -434,6 +434,7 @@ func (c *client) GetPodFlowKeys(interfaceName string) []string {
 func (c *client) InstallServiceGroup(groupID binding.GroupIDType, withSessionAffinity bool, endpoints []proxy.Endpoint) error {
 	c.replayMutex.RLock()
 	defer c.replayMutex.RUnlock()
+
 	group := c.serviceEndpointGroup(groupID, withSessionAffinity, endpoints...)
 	if err := group.Add(); err != nil {
 		return fmt.Errorf("error when installing Service Endpoints Group: %w", err)
@@ -497,7 +498,7 @@ func (c *client) InstallServiceFlows(groupID binding.GroupIDType, svcIP net.IP, 
 	c.replayMutex.RLock()
 	defer c.replayMutex.RUnlock()
 	var flows []binding.Flow
-	flows = append(flows, c.serviceLBFlow(groupID, svcIP, svcPort, protocol))
+	flows = append(flows, c.serviceLBFlow(groupID, svcIP, svcPort, protocol, affinityTimeout != 0))
 	if affinityTimeout != 0 {
 		flows = append(flows, c.serviceLearnFlow(groupID, svcIP, svcPort, protocol, affinityTimeout))
 	}

--- a/pkg/agent/openflow/pipeline.go
+++ b/pkg/agent/openflow/pipeline.go
@@ -1903,12 +1903,21 @@ func (c *client) serviceLearnFlow(groupID binding.GroupIDType, svcIP net.IP, svc
 
 // serviceLBFlow generates the flow which uses the specific group to do Endpoint
 // selection.
-func (c *client) serviceLBFlow(groupID binding.GroupIDType, svcIP net.IP, svcPort uint16, protocol binding.Protocol) binding.Flow {
+func (c *client) serviceLBFlow(groupID binding.GroupIDType, svcIP net.IP, svcPort uint16, protocol binding.Protocol, withSessionAffinity bool) binding.Flow {
+	var lbResultMark uint32
+	if withSessionAffinity {
+		lbResultMark = marksRegServiceNeedLearn
+	} else {
+		lbResultMark = marksRegServiceSelected
+	}
+
 	return c.pipeline[serviceLBTable].BuildFlow(priorityNormal).
 		MatchProtocol(protocol).
 		MatchDstPort(svcPort, nil).
 		MatchDstIP(svcIP).
 		MatchRegRange(int(serviceLearnReg), marksRegServiceNeedLB, serviceLearnRegRange).
+		Action().LoadRegRange(int(serviceLearnReg), lbResultMark, serviceLearnRegRange).
+		Action().LoadRegRange(int(marksReg), macRewriteMark, macRewriteMarkRange).
 		Action().Group(groupID).
 		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
 		Done()
@@ -1974,13 +1983,10 @@ func (c *client) hairpinSNATFlow(endpointIP net.IP) binding.Flow {
 func (c *client) serviceEndpointGroup(groupID binding.GroupIDType, withSessionAffinity bool, endpoints ...proxy.Endpoint) binding.Group {
 	group := c.bridge.CreateGroup(groupID).ResetBuckets()
 	var resubmitTableID binding.TableIDType
-	var lbResultMark uint32
 	if withSessionAffinity {
 		resubmitTableID = serviceLBTable
-		lbResultMark = marksRegServiceNeedLearn
 	} else {
 		resubmitTableID = endpointDNATTable
-		lbResultMark = marksRegServiceSelected
 	}
 
 	for _, endpoint := range endpoints {
@@ -1993,8 +1999,6 @@ func (c *client) serviceEndpointGroup(groupID binding.GroupIDType, withSessionAf
 			group = group.Bucket().Weight(100).
 				LoadReg(int(endpointIPReg), ipVal).
 				LoadRegRange(int(endpointPortReg), uint32(portVal), endpointPortRegRange).
-				LoadRegRange(int(serviceLearnReg), lbResultMark, serviceLearnRegRange).
-				LoadRegRange(int(marksReg), macRewriteMark, macRewriteMarkRange).
 				ResubmitToTable(resubmitTableID).
 				Done()
 		} else if ipProtocol == binding.ProtocolIPv6 {
@@ -2002,8 +2006,6 @@ func (c *client) serviceEndpointGroup(groupID binding.GroupIDType, withSessionAf
 			group = group.Bucket().Weight(100).
 				LoadXXReg(int(endpointIPv6XXReg), ipVal).
 				LoadRegRange(int(endpointPortReg), uint32(portVal), endpointPortRegRange).
-				LoadRegRange(int(serviceLearnReg), lbResultMark, serviceLearnRegRange).
-				LoadRegRange(int(marksReg), macRewriteMark, macRewriteMarkRange).
 				ResubmitToTable(resubmitTableID).
 				Done()
 		}

--- a/pkg/agent/proxy/endpoints.go
+++ b/pkg/agent/proxy/endpoints.go
@@ -108,7 +108,7 @@ func (t *endpointsChangesTracker) OnEndpointUpdate(previous, current *corev1.End
 	return len(t.changes) > 0
 }
 
-// EndpointSliceUpdate updates the given service's endpoints change map based on the <previous, current> endpoints pair.
+// OnEndpointSliceUpdate updates the given service's endpoints change map based on the <previous, current> endpoints pair.
 // It returns true if items changed, otherwise it returns false. Will add/update/delete items of endpointsChange Map.
 // If removeSlice is true, slice will be removed, otherwise it will be added or updated.
 func (t *endpointsChangesTracker) OnEndpointSliceUpdate(endpointSlice *discovery.EndpointSlice, removeSlice bool) bool {

--- a/pkg/agent/proxy/endpoints.go
+++ b/pkg/agent/proxy/endpoints.go
@@ -220,3 +220,16 @@ func (t *endpointsChangesTracker) Update(em types.EndpointsMap) {
 		}
 	}
 }
+
+// byEndpoint helps sort Endpoint
+type byEndpoint []k8sproxy.Endpoint
+
+func (p byEndpoint) Len() int {
+	return len(p)
+}
+func (p byEndpoint) Swap(i, j int) {
+	p[i], p[j] = p[j], p[i]
+}
+func (p byEndpoint) Less(i, j int) bool {
+	return p[i].String() < p[j].String()
+}

--- a/pkg/agent/proxy/proxier.go
+++ b/pkg/agent/proxy/proxier.go
@@ -17,6 +17,7 @@ package proxy
 import (
 	"fmt"
 	"net"
+	"sort"
 	"strings"
 	"sync"
 	"time"
@@ -124,8 +125,8 @@ func (p *proxier) removeStaleServices() {
 		}
 		svcInfo := svcPort.(*types.ServiceInfo)
 		klog.V(2).Infof("Removing stale Service: %s %s", svcPortName.Name, svcInfo.String())
-		if p.oversizeServiceSet.Has(svcInfo.String()) {
-			p.oversizeServiceSet.Delete(svcInfo.String())
+		if p.oversizeServiceSet.Has(svcPortName.String()) {
+			p.oversizeServiceSet.Delete(svcPortName.String())
 		}
 		if err := p.ofClient.UninstallServiceFlows(svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol); err != nil {
 			klog.Errorf("Failed to remove flows of Service %v: %v", svcPortName, err)
@@ -274,6 +275,34 @@ func (p *proxier) installServices() {
 			}
 			endpointUpdateList = append(endpointUpdateList, endpoint)
 		}
+
+		// If the length of endpointUpdateList > maxEndpoints, endpointUpdateList should be cut. However, the iteration
+		// of map in Golang is random, so endpointUpdateList are not always in the same order. If endpointUpdateList is
+		// cut directly without any sorting, some Endpoints may not be installed on the cut endpointUpdateList(Last sync,
+		// these Endpoints may be cut off from endpointUpdateList, they are of course not installed.) So cutting
+		// endpointUpdateList after sorting can avoid this situation in some degree.
+		if len(endpointUpdateList) > maxEndpoints {
+			if !p.oversizeServiceSet.Has(svcPortName.String()) {
+				klog.Warningf("Since Endpoints of Service %s exceeds %d, extra Endpoints will be dropped", svcPortName.String(), maxEndpoints)
+				p.oversizeServiceSet.Insert(svcPortName.String())
+			}
+
+			needUpdateEndpoints = false
+			sort.Sort(byEndpoint(endpointUpdateList))
+			endpointUpdateList = endpointUpdateList[:maxEndpoints]
+			// Check if the cut endpointUpdateList are all installed.
+			for _, endpoint := range endpointUpdateList {
+				if _, ok := endpointsInstalled[endpoint.String()]; !ok {
+					needUpdateEndpoints = true
+					break
+				}
+			}
+		} else {
+			if p.oversizeServiceSet.Has(svcPortName.String()) {
+				p.oversizeServiceSet.Delete(svcPortName.String())
+			}
+		}
+
 		if len(endpoints) < len(endpointsInstalled) { // There are Endpoints which expired.
 			klog.V(2).Infof("Some Endpoints of Service %s removed, updating Endpoints", svcInfo.String())
 			needUpdateEndpoints = true
@@ -303,18 +332,6 @@ func (p *proxier) installServices() {
 		}
 
 		if needUpdateEndpoints {
-			if len(endpointUpdateList) > maxEndpoints {
-				endpointUpdateList = endpointUpdateList[:maxEndpoints]
-				if !p.oversizeServiceSet.Has(svcInfo.String()) {
-					klog.Warningf("Since Endpoints of Service %s exceeds %d, extra Endpoints will be dropped.", svcInfo.String(), maxEndpoints)
-					p.oversizeServiceSet.Insert(svcInfo.String())
-				}
-			} else {
-				if p.oversizeServiceSet.Has(svcInfo.String()) {
-					p.oversizeServiceSet.Delete(svcInfo.String())
-				}
-			}
-
 			err := p.ofClient.InstallEndpointFlows(svcInfo.OFProtocol, endpointUpdateList)
 			if err != nil {
 				klog.Errorf("Error when installing Endpoints flows: %v", err)

--- a/pkg/agent/proxy/proxier.go
+++ b/pkg/agent/proxy/proxier.go
@@ -269,37 +269,37 @@ func (p *proxier) installServices() {
 		}
 
 		var endpointUpdateList []k8sproxy.Endpoint
-		for _, endpoint := range endpoints { // Check if there is any installed Endpoint which is not expected anymore.
-			if _, ok := endpointsInstalled[endpoint.String()]; !ok { // There is an expected Endpoint which is not installed.
-				needUpdateEndpoints = true
-			}
-			endpointUpdateList = append(endpointUpdateList, endpoint)
-		}
-
-		// If the length of endpointUpdateList > maxEndpoints, endpointUpdateList should be cut. However, the iteration
-		// of map in Golang is random, so endpointUpdateList are not always in the same order. If endpointUpdateList is
-		// cut directly without any sorting, some Endpoints may not be installed on the cut endpointUpdateList(Last sync,
-		// these Endpoints may be cut off from endpointUpdateList, they are of course not installed.) So cutting
-		// endpointUpdateList after sorting can avoid this situation in some degree.
-		if len(endpointUpdateList) > maxEndpoints {
+		if len(endpoints) > maxEndpoints {
 			if !p.oversizeServiceSet.Has(svcPortName.String()) {
 				klog.Warningf("Since Endpoints of Service %s exceeds %d, extra Endpoints will be dropped", svcPortName.String(), maxEndpoints)
 				p.oversizeServiceSet.Insert(svcPortName.String())
 			}
+			// If the length of endpoints > maxEndpoints, endpoints should be cut. However, endpoints is a map. Therefore,
+			// iterate the map and append every Endpoint to a slice endpointList. Since the iteration order of map in
+			// Golang is random, if cut directly without any sorting, some Endpoints may not be installed. So cutting
+			// slice endpointList after sorting can avoid this situation in some degree.
+			var endpointList []k8sproxy.Endpoint
+			for _, endpoint := range endpoints {
+				endpointList = append(endpointList, endpoint)
+			}
+			sort.Sort(byEndpoint(endpointList))
+			endpointList = endpointList[:maxEndpoints]
 
-			needUpdateEndpoints = false
-			sort.Sort(byEndpoint(endpointUpdateList))
-			endpointUpdateList = endpointUpdateList[:maxEndpoints]
-			// Check if the cut endpointUpdateList are all installed.
-			for _, endpoint := range endpointUpdateList {
-				if _, ok := endpointsInstalled[endpoint.String()]; !ok {
+			for _, endpoint := range endpointList { // Check if there is any installed Endpoint which is not expected anymore.
+				if _, ok := endpointsInstalled[endpoint.String()]; !ok { // There is an expected Endpoint which is not installed.
 					needUpdateEndpoints = true
-					break
 				}
+				endpointUpdateList = append(endpointUpdateList, endpoint)
 			}
 		} else {
 			if p.oversizeServiceSet.Has(svcPortName.String()) {
 				p.oversizeServiceSet.Delete(svcPortName.String())
+			}
+			for _, endpoint := range endpoints { // Check if there is any installed Endpoint which is not expected anymore.
+				if _, ok := endpointsInstalled[endpoint.String()]; !ok { // There is an expected Endpoint which is not installed.
+					needUpdateEndpoints = true
+				}
+				endpointUpdateList = append(endpointUpdateList, endpoint)
 			}
 		}
 

--- a/pkg/agent/proxy/proxier.go
+++ b/pkg/agent/proxy/proxier.go
@@ -25,6 +25,7 @@ import (
 	"k8s.io/api/discovery/v1beta1"
 	"k8s.io/apimachinery/pkg/runtime"
 	k8sapitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/tools/record"
 	"k8s.io/klog"
@@ -42,6 +43,10 @@ import (
 const (
 	resyncPeriod  = time.Minute
 	componentName = "antrea-agent-proxy"
+	// Due to the maximum message size in Openflow 1.3 and the implementation of Services in Antrea, the maximum number
+	// of Endpoints that Antrea can support at the moment is 800. If the number of Endpoints for a given Service exceeds
+	// 800, extra Endpoints will be dropped.
+	maxEndpoints = 800
 )
 
 // Proxier wraps proxy.Provider and adds extra methods. It is introduced for
@@ -90,6 +95,8 @@ type proxier struct {
 	serviceStringMap map[string]k8sproxy.ServicePortName
 	// serviceStringMapMutex protects serviceStringMap object.
 	serviceStringMapMutex sync.Mutex
+	// oversizeServiceSet records the Services that have more than 800 Endpoints.
+	oversizeServiceSet sets.String
 
 	runner              *k8sproxy.BoundedFrequencyRunner
 	stopChan            <-chan struct{}
@@ -117,6 +124,9 @@ func (p *proxier) removeStaleServices() {
 		}
 		svcInfo := svcPort.(*types.ServiceInfo)
 		klog.V(2).Infof("Removing stale Service: %s %s", svcPortName.Name, svcInfo.String())
+		if p.oversizeServiceSet.Has(svcInfo.String()) {
+			p.oversizeServiceSet.Delete(svcInfo.String())
+		}
 		if err := p.ofClient.UninstallServiceFlows(svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol); err != nil {
 			klog.Errorf("Failed to remove flows of Service %v: %v", svcPortName, err)
 			continue
@@ -293,6 +303,18 @@ func (p *proxier) installServices() {
 		}
 
 		if needUpdateEndpoints {
+			if len(endpointUpdateList) > maxEndpoints {
+				endpointUpdateList = endpointUpdateList[:maxEndpoints]
+				if !p.oversizeServiceSet.Has(svcInfo.String()) {
+					klog.Warningf("Since Endpoints of Service %s exceeds %d, extra Endpoints will be dropped.", svcInfo.String(), maxEndpoints)
+					p.oversizeServiceSet.Insert(svcInfo.String())
+				}
+			} else {
+				if p.oversizeServiceSet.Has(svcInfo.String()) {
+					p.oversizeServiceSet.Delete(svcInfo.String())
+				}
+			}
+
 			err := p.ofClient.InstallEndpointFlows(svcInfo.OFProtocol, endpointUpdateList)
 			if err != nil {
 				klog.Errorf("Error when installing Endpoints flows: %v", err)
@@ -605,6 +627,7 @@ func NewProxier(
 		endpointsMap:             types.EndpointsMap{},
 		endpointReferenceCounter: map[string]int{},
 		serviceStringMap:         map[string]k8sproxy.ServicePortName{},
+		oversizeServiceSet:       sets.NewString(),
 		groupCounter:             types.NewGroupCounter(),
 		ofClient:                 ofClient,
 		isIPv6:                   isIPv6,

--- a/test/e2e/proxy_test.go
+++ b/test/e2e/proxy_test.go
@@ -281,9 +281,9 @@ func testProxyServiceLifeCycle(ipFamily *corev1.IPFamily, ingressIPs []string, d
 
 	var groupKeyword string
 	if *ipFamily == corev1.IPv6Protocol {
-		groupKeyword = fmt.Sprintf("set_field:0x%s->xxreg3,load:0x%x->NXM_NX_REG4[0..15],load:0x2->NXM_NX_REG4[16..18]", strings.TrimLeft(hex.EncodeToString(nginxIPs.ipv6.To16()), "0"), 80)
+		groupKeyword = fmt.Sprintf("set_field:0x%s->xxreg3,load:0x%x->NXM_NX_REG4[0..15]", strings.TrimLeft(hex.EncodeToString(nginxIPs.ipv6.To16()), "0"), 80)
 	} else {
-		groupKeyword = fmt.Sprintf("load:0x%s->NXM_NX_REG3[],load:0x%x->NXM_NX_REG4[0..15],load:0x2->NXM_NX_REG4[16..18]", strings.TrimLeft(hex.EncodeToString(nginxIPs.ipv4.To4()), "0"), 80)
+		groupKeyword = fmt.Sprintf("load:0x%s->NXM_NX_REG3[],load:0x%x->NXM_NX_REG4[0..15]", strings.TrimLeft(hex.EncodeToString(nginxIPs.ipv4.To4()), "0"), 80)
 	}
 	groupOutput, _, err := data.runCommandFromPod(metav1.NamespaceSystem, agentName, "antrea-agent", []string{"ovs-ofctl", "dump-groups", defaultBridgeName})
 	require.NoError(t, err)

--- a/test/integration/agent/openflow_test.go
+++ b/test/integration/agent/openflow_test.go
@@ -600,11 +600,16 @@ func expectedProxyServiceGroupAndFlows(gid uint32, svc svcConfig, endpointList [
 		nw_proto = 132
 		learnProtoField = "OXM_OF_SCTP_DST[]"
 	}
+
+	serviceLearnReg := 2
+	if stickyAge != 0 {
+		serviceLearnReg = 3
+	}
 	cookieAllocator := cookie.NewAllocator(roundInfo.RoundNum)
 	svcFlows := expectTableFlows{tableID: 41, flows: []*ofTestUtils.ExpectFlow{
 		{
 			MatchStr: fmt.Sprintf("priority=200,%s,reg4=0x10000/0x70000,nw_dst=%s,tp_dst=%d", string(svc.protocol), svc.ip.String(), svc.port),
-			ActStr:   fmt.Sprintf("group:%d", gid),
+			ActStr:   fmt.Sprintf("load:0x%x->NXM_NX_REG4[16..18],load:0x1->NXM_NX_REG0[19],group:%d", serviceLearnReg, gid),
 		},
 		{
 			MatchStr: fmt.Sprintf("priority=190,%s,reg4=0x30000/0x70000,nw_dst=%s,tp_dst=%d", string(svc.protocol), svc.ip.String(), svc.port),
@@ -617,7 +622,7 @@ func expectedProxyServiceGroupAndFlows(gid uint32, svc svcConfig, endpointList [
 	for _, ep := range endpointList {
 		epIP := ipToHexString(net.ParseIP(ep.IP()))
 		epPort, _ := ep.Port()
-		bucket := fmt.Sprintf("weight:100,actions=load:%s->NXM_NX_REG3[],load:0x%x->NXM_NX_REG4[0..15],load:0x2->NXM_NX_REG4[16..18],load:0x1->NXM_NX_REG0[19],resubmit(,42)", epIP, epPort)
+		bucket := fmt.Sprintf("weight:100,actions=load:%s->NXM_NX_REG3[],load:0x%x->NXM_NX_REG4[0..15],resubmit(,42)", epIP, epPort)
 		groupBuckets = append(groupBuckets, bucket)
 
 		unionVal := (0b010 << 16) + uint32(epPort)

